lib/improver/generate_ancillaries/generate_topographic_zone_weights.py

```diff
@@ -0,0 +1,277 @@
+            orography.units)
+
+        # Read bands from cube, now that they can be guaranteed to be in the
+        # same units as the orography. The bands are converted to a list, so
+        # that they can be iterated through.
+        bands = list(topographic_zone_weights.coord("topographic_zone").bounds)
+        midpoints = topographic_zone_weights.coord("topographic_zone").points
+
+        # Raise a warning, if orography extremes are outside the extremes of
+        # the bands.
+        if np.max(orography.data) > np.max(bands):
+            msg = ("The maximum orography is greater than the uppermost band. "
+                   "This will potentially cause the topographic zone weights "
+                   "to not sum to 1 for a given grid point.")
+            warnings.warn(msg)
+
+        if np.min(orography.data) < np.min(bands):
+            msg = ("The minimum orography is lower than the lowest band. "
+                   "This will potentially cause the topographic zone weights "
+                   "to not sum to 1 for a given grid point.")
+            warnings.warn(msg)
+
+        # Insert the appropriate weights into the topographic zone cube. This
+        # includes the weights from the band that a point is in, as well as
+        # the contribution from an adjacent band.
+        for band_number, band in enumerate(bands):
+            # Determine the points that are within the specified band.
+            mask_y, mask_x = (
+                np.where((orography.data > band[0]) &
+                         (orography.data <= band[1])))
+            orography_band = np.full(orography.shape, np.nan)
+            orography_band[mask_y, mask_x] = orography.data[mask_y, mask_x]
+
+            # Calculate the weights. This involves calculating the
+            # weights for all the orography but only inserting weights
+            # that are within the band into the topographic_zone_weights cube.
+            weights = self.calculate_weights(orography_band, band)
+            topographic_zone_weights.data[band_number, mask_y, mask_x] = (
+                weights[mask_y, mask_x])
+
+            # Calculate the contribution to the weights from the adjacent
+            # lower band.
+            topographic_zone_weights.data = (
+                self.add_weight_to_lower_adjacent_band(
+                    topographic_zone_weights.data, orography_band,
+                    midpoints[band_number], band_number))
+
+            # Calculate the contribution to the weights from the adjacent
+            # upper band.
+            topographic_zone_weights.data = (
+                self.add_weight_to_upper_adjacent_band(
+                    topographic_zone_weights.data, orography_band,
+                    midpoints[band_number], band_number,
+                    len(bands)-1))
+
+        # Metadata updates
+        topographic_zone_weights.rename("Topographic_zone_weights")
```

Comment:
Just noticed looking at the acceptance test data that we don't normally have capital letters in cube names.